drivers/serial/uart_nrf5e.c

```diff
@@ -0,0 +1,873 @@
+	case 19200:
+		divisor = NRF5_UART_BAUDRATE_19200;
+		break;
+	case 28800:
+		divisor = NRF5_UART_BAUDRATE_28800;
+		break;
+	case 38400:
+		divisor = NRF5_UART_BAUDRATE_38400;
+		break;
+	case 57600:
+		divisor = NRF5_UART_BAUDRATE_57600;
+		break;
+	case 76800:
+		divisor = NRF5_UART_BAUDRATE_76800;
+		break;
+	case 115200:
+		divisor = NRF5_UART_BAUDRATE_115200;
+		break;
+	case 230400:
+		divisor = NRF5_UART_BAUDRATE_230400;
+		break;
+	case 250000:
+		divisor = NRF5_UART_BAUDRATE_250000;
+		break;
+	case 460800:
+		divisor = NRF5_UART_BAUDRATE_460800;
+		break;
+	case 921600:
+		divisor = NRF5_UART_BAUDRATE_921600;
+		break;
+	case 1000000:
+		divisor = NRF5_UART_BAUDRATE_1000000;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	uart->BAUDRATE = divisor << UART_BAUDRATE_BAUDRATE_Pos;
+
+	return 0;
+}
+
+/**
+ * @brief Initialize UART channel
+ *
+ * This routine is called to reset the chip in a quiescent state.
+ * It is assumed that this function is called only once per UART.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0 on success
+ */
+static int uart_nrf5e_init(struct device *dev)
+{
+	volatile struct _uart *uart = UART_STRUCT(dev);
+	struct device *gpio_dev;
+	int err;
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_TX_PIN,
+				  (GPIO_DIR_OUT | GPIO_PUD_PULL_UP));
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_RX_PIN,
+				  (GPIO_DIR_IN));
+
+	uart->PSELTXD = CONFIG_UART_NRF5E_GPIO_TX_PIN;
+	uart->PSELRXD = CONFIG_UART_NRF5E_GPIO_RX_PIN;
+
+#ifdef CONFIG_UART_NRF5E_FLOW_CONTROL
+
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_RTS_PIN,
+				  (GPIO_DIR_OUT | GPIO_PUD_PULL_UP));
+	(void) gpio_pin_configure(gpio_dev,
+				  CONFIG_UART_NRF5E_GPIO_CTS_PIN,
+				  (GPIO_DIR_IN));
+
+	uart->PSELRTS = CONFIG_UART_NRF5E_GPIO_RTS_PIN;
+	uart->PSELCTS = CONFIG_UART_NRF5E_GPIO_CTS_PIN;
+	uart->CONFIG = (UART_CONFIG_HWFC_Enabled << UART_CONFIG_HWFC_Pos);
+
+#endif /* CONFIG_UART_NRF5E_FLOW_CONTROL */
+
+	DEV_DATA(dev)->baud_rate = CONFIG_UART_NRF5_BAUD_RATE;
+
+	/* Set baud rate */
+	err = baudrate_set(dev, DEV_DATA(dev)->baud_rate,
+		  DEV_CFG(dev)->sys_clk_freq);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	uart->ENABLE = (UARTE_ENABLE_ENABLE_Enabled << UARTE_ENABLE_ENABLE_Pos);
+
+	uart->EVENTS_ENDTX = 0;
+	uart->EVENTS_ENDRX = 0;
+	uart->EVENTS_RXDRDY = 0;
+
+	dev->driver_api = &uart_nrf5e_driver_api;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	DEV_CFG(dev)->irq_config_func(dev);
+	uart_nrf5e_mask(dev, 0);
+	uart_nrf5e_fifo_rx_setup(dev);
+#endif
+
+	return 0;
+}
+
+/**
+ * @brief Poll the device for input.
+ *
+ * @param dev UART device struct
+ * @param c Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+
+static int uart_nrf5e_poll_in(struct device *dev, unsigned char *c)
+{
+
+#ifndef CONFIG_UART_INTERRUPT_DRIVEN
```

Comment:
I have some concerns for the non-interrupt-driven case:
1. For an RXDRDY event to be non-zero (line 345), the UARTE receiver needs to have been started. Is this occurring, anywhere, for the non-IRQ-driven case?
2. In any case, the code after line 344 assumes that a character is received already. Then why does the do-while block in lines 350-359 try to re-start RXing?